[JSC] Make BaselineCallLinkInfo more compact

Constellation opened the pull request:
#### 349d9d98217583326815bc41d75d96f16bb47c0b
<pre>
[JSC] Make BaselineCallLinkInfo more compact
<a href="https://bugs.webkit.org/show_bug.cgi?id=248704">https://bugs.webkit.org/show_bug.cgi?id=248704</a>
rdar://102929204

Reviewed by Mark Lam.

This patch makes BaselineCallLinkInfo more compact by reducing size of m_maxArgumentCountIncludingThis from unsigned to uint8_t.
This is OK since this data is profiling information and only meaningful value is under 100 (since varargs call inlining threshold
is 100). We also store BytecodeIndex for BaselineCallLinkInfo instead of CodeOrigin. By doing them, we can make
sizeof(BaselineCallLinkInfo) from 88 to 80 with the other reordering.

* Source/JavaScriptCore/bytecode/CallLinkInfo.cpp:
(JSC::OptimizingCallLinkInfo::setDirectCallMaxArgumentCountIncludingThis):
(JSC::CallLinkInfo::visitWeak):
(JSC::BaselineCallLinkInfo::initialize):
(JSC::CallLinkInfo::setMaxArgumentCountIncludingThis): Deleted.
* Source/JavaScriptCore/bytecode/CallLinkInfo.h:
(JSC::CallLinkInfo::offsetOfMaxArgumentCountIncludingThisForVarargs):
(JSC::CallLinkInfo::maxArgumentCountIncludingThisForVarargs):
(JSC::CallLinkInfo::updateMaxArgumentCountIncludingThisForVarargs):
(JSC::CallLinkInfo::CallLinkInfo):
(JSC::CallLinkInfo::codeOrigin const):
(JSC::CallLinkInfo::offsetOfMaxArgumentCountIncludingThis): Deleted.
(JSC::CallLinkInfo::maxArgumentCountIncludingThis): Deleted.
(JSC::CallLinkInfo::updateMaxArgumentCountIncludingThis): Deleted.
(JSC::CallLinkInfo::codeOrigin): Deleted.
(JSC::CallLinkInfo::setCallLinkInfoGPR): Deleted.
* Source/JavaScriptCore/bytecode/CallLinkStatus.cpp:
(JSC::CallLinkStatus::computeFor):
(JSC::CallLinkStatus::dump const):
* Source/JavaScriptCore/bytecode/CallLinkStatus.h:
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleVarargsInlining):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::emitCall):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::emitCall):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/jit/JITCall.cpp:
(JSC::JIT::compileSetupFrame):
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::varargsSetup):

Canonical link: <a href="https://commits.webkit.org/257325@main">https://commits.webkit.org/257325@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d12043077110dd21c8a1b4729ab982726752975a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98592 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7802 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31718 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108014 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168276 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8313 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85180 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91127 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/105988 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104274 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89854 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33295 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88117 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76227 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/89382 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1735 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/22741 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/85152 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1646 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28737 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6586 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42175 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/88004 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2529 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3035 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19714 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/bytecode/CallLinkInfo.h:
```diff
@@ -527,13 +522,19 @@ class OptimizingCallLinkInfo final : public CallLinkInfo {
-    CodeLocationNearCall<JSInternalPtrTag> m_callLocation;
+    CodeOrigin m_codeOrigin;
+    CodeLocationNearCall<JSInternalPtrTag> m_callLocation NO_UNIQUE_ADDRESS;
+    GPRReg m_calleeGPR { InvalidGPRReg };
+    GPRReg m_callLinkInfoGPR { InvalidGPRReg };
```
These `GPRReg` are `uint8_t`.  Are there savings to be had if we move them above `m_codeOrigin`?

Constellation replied:
Not much. We attached NO_UNIQUE_ADDRESS so that we are putting them into the padding.

MenloDorian reviewed (approved):
r=me

Constellation committed:
[JSC] Make BaselineCallLinkInfo more compact
https://bugs.webkit.org/show_bug.cgi?id=248704
rdar://102929204

Reviewed by Mark Lam.

This patch makes BaselineCallLinkInfo more compact by reducing size of m_maxArgumentCountIncludingThis from unsigned to uint8_t.
This is OK since this data is profiling information and only meaningful value is under 100 (since varargs call inlining threshold
is 100). We also store BytecodeIndex for BaselineCallLinkInfo instead of CodeOrigin. By doing them, we can make
sizeof(BaselineCallLinkInfo) from 88 to 80 with the other reordering.

* Source/JavaScriptCore/bytecode/CallLinkInfo.cpp:
(JSC::OptimizingCallLinkInfo::setDirectCallMaxArgumentCountIncludingThis):
(JSC::CallLinkInfo::visitWeak):
(JSC::BaselineCallLinkInfo::initialize):
(JSC::CallLinkInfo::setMaxArgumentCountIncludingThis): Deleted.
* Source/JavaScriptCore/bytecode/CallLinkInfo.h:
(JSC::CallLinkInfo::offsetOfMaxArgumentCountIncludingThisForVarargs):
(JSC::CallLinkInfo::maxArgumentCountIncludingThisForVarargs):
(JSC::CallLinkInfo::updateMaxArgumentCountIncludingThisForVarargs):
(JSC::CallLinkInfo::CallLinkInfo):
(JSC::CallLinkInfo::codeOrigin const):
(JSC::CallLinkInfo::offsetOfMaxArgumentCountIncludingThis): Deleted.
(JSC::CallLinkInfo::maxArgumentCountIncludingThis): Deleted.
(JSC::CallLinkInfo::updateMaxArgumentCountIncludingThis): Deleted.
(JSC::CallLinkInfo::codeOrigin): Deleted.
(JSC::CallLinkInfo::setCallLinkInfoGPR): Deleted.
* Source/JavaScriptCore/bytecode/CallLinkStatus.cpp:
(JSC::CallLinkStatus::computeFor):
(JSC::CallLinkStatus::dump const):
* Source/JavaScriptCore/bytecode/CallLinkStatus.h:
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleVarargsInlining):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::emitCall):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::emitCall):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/jit/JITCall.cpp:
(JSC::JIT::compileSetupFrame):
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::varargsSetup):

Canonical link: https://commits.webkit.org/257325@main
